pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -336,63 +335,59 @@ func (e *EndpointController) syncService(key string) error {
-			if err != nil {
-				glog.V(4).Infof("Failed to find port for service %s/%s: %v", service.Namespace, service.Name, err)
-				continue
-			}
-			if len(pod.Status.PodIP) == 0 {
-				glog.V(5).Infof("Failed to find an IP for pod %s/%s", pod.Namespace, pod.Name)
-				continue
-			}
-			if !tolerateUnreadyEndpoints && pod.DeletionTimestamp != nil {
-				glog.V(5).Infof("Pod is being deleted %s/%s", pod.Namespace, pod.Name)
-				continue
-			}
+		epa := v1.EndpointAddress{
+			IP:       pod.Status.PodIP,
+			NodeName: &pod.Spec.NodeName,
+			TargetRef: &v1.ObjectReference{
+				Kind:            "Pod",
+				Namespace:       pod.ObjectMeta.Namespace,
+				Name:            pod.ObjectMeta.Name,
+				UID:             pod.ObjectMeta.UID,
+				ResourceVersion: pod.ObjectMeta.ResourceVersion,
+			}}
+
+		hostname := pod.Spec.Hostname
+		if len(hostname) > 0 && pod.Spec.Subdomain == service.Name && service.Namespace == pod.Namespace {
+			epa.Hostname = hostname
+		}
 
-			epp := v1.EndpointPort{Name: portName, Port: int32(portNum), Protocol: portProto}
-			epa := v1.EndpointAddress{
-				IP:       pod.Status.PodIP,
-				NodeName: &pod.Spec.NodeName,
-				TargetRef: &v1.ObjectReference{
-					Kind:            "Pod",
-					Namespace:       pod.ObjectMeta.Namespace,
-					Name:            pod.ObjectMeta.Name,
-					UID:             pod.ObjectMeta.UID,
-					ResourceVersion: pod.ObjectMeta.ResourceVersion,
-				}}
-
-			hostname := pod.Spec.Hostname
-			if len(hostname) > 0 &&
-				pod.Spec.Subdomain == service.Name &&
-				service.Namespace == pod.Namespace {
-				epa.Hostname = hostname
+		// Allow headless service not to have ports.
+		if len(service.Spec.Ports) == 0 {
+			if service.Spec.ClusterIP == api.ClusterIPNone {
+				epp := v1.EndpointPort{Port: 0, Protocol: v1.ProtocolTCP}
+				subsets, readyEps, notReadyEps = addEndpointSubset(subsets, pod, epa, epp, tolerateUnreadyEndpoints)
 			}
-
-			if tolerateUnreadyEndpoints || podutil.IsPodReady(pod) {
-				subsets = append(subsets, v1.EndpointSubset{
-					Addresses: []v1.EndpointAddress{epa},
-					Ports:     []v1.EndpointPort{epp},
-				})
-				readyEps++
-			} else {
-				glog.V(5).Infof("Pod is out of service: %v/%v", pod.Namespace, pod.Name)
-				subsets = append(subsets, v1.EndpointSubset{
-					NotReadyAddresses: []v1.EndpointAddress{epa},
-					Ports:             []v1.EndpointPort{epp},
-				})
-				notReadyEps++
+		} else {
+			for i := range service.Spec.Ports {
+				servicePort := &service.Spec.Ports[i]
+
+				portName := servicePort.Name
+				portProto := servicePort.Protocol
+				portNum, err := podutil.FindPort(pod, servicePort)
+				if err != nil {
+					glog.V(4).Infof("Failed to find port for service %s/%s: %v", service.Namespace, service.Name, err)
+					continue
+				}
+
+				epp := v1.EndpointPort{Name: portName, Port: int32(portNum), Protocol: portProto}
+				subsets, readyEps, notReadyEps = addEndpointSubset(subsets, pod, epa, epp, tolerateUnreadyEndpoints)
```

Comment:
you want to aggregate `readyEps` and `notReadyEps`, right? 
Otherwise it will always ended up with 1 or 0.